### ISSUE TITLE
Alternative Approaches to Builtin Tag Naming

### ISSUE THREAD

**Issue by catmando** [hyperstack-org/hyperstack#147]
Currently builtin tags are UPCASED.   Some folks find this "ugly" and raises 3 questions:

1.  Is it really Ugly?  Or is it just *different* and a bit unexpected?
2. Is it a significant barrier to people accepting Hyperstack?
3. And most important:  Is there a better way to do it?

(see https://hyperstack.org/ for some examples)

Here are the challenges:

There about 200 standard HTML and SVG tags.  These includes tags like `param`, `p` and `map` that conflict with existing methods; tags with simple names like `i`, `a`, and `b` that will conflict with names often used as local variables; and tags with common names that people may use in their code like `label`, `text`, `time`, `object`. 

The problem with these conflicts is that it can result in unexpected or hard to debug behavior.

For example, a large component class may define an instance method `label`.  Then later in the code somebody forgets and tries to use `label` as an HTML tag.  Or the method call is named `labels`, but is misspelled `label`.

The other problem is simply that with 200 tags, its can be hard to distinguish tags especially little used tags, from the rest of the code. 

These are real problems that early adopters encountered in the early days of Hyperstack.

**What we need**

Please think on these issues, and comment on how you feel about them, and any potential solutions to problem that you feel would be agreeable or at least better.

Thanks!

**Comment by galtenberg** [comment]
Foreign at first, but doesn't seem like a blocker after a day of using.

The issue may be more of the mix/match with ruby keywords/syntax around tag code (`render`, `end`, `class PascalCase`) plus (parens) and {braces} - pretty chaotic initially.

Did you try with symbols? (`:div`, `:i`, `:span`, etc) Not sure that's compatible with method definitions and other primary uses.

**Comment by catmando** [comment]
@galtenberg thanks for the feedback.  The problem with symbols as I see it is that you would have to do something like `tag(:div) { 'my div' }` or `:div.tag { ' my div ' }`

even if did something really short like `:div.i!` (where i! stands for render) it adds 4 characters per tag, and really looks pretty ugly.

**Comment by mwhobrey** [comment]
I don't think it is ugly, just different. By convention, I (and a lot of people) usually reserve all caps as constants, but I really don't have a problem translating HS tags as constants (in my brain). It was a bit shocking at first but after I made the connection (to me) that thinking of the tags as a kind of a functional type constant it really clicked and seemed to make more sense, again, to me. 

The more I have used HS the more I have grown accustomed to the syntax, and even if the style is something that is offputting to some, I think that a style annoyance is a minimal tradeoff for the other amazingness with HS, IMHO.

Out of all the other ways to implement the tags, I think this one has the least amount of drawbacks (aside from its potential reception). Most, if not all, other options are going to add a lot more keystrokes and someone somewhere will find a reason to complain.

**Comment by barriehadfield** [comment]
The bit I have a  hump with is the conversion of `param`s from snake_case to CamelCase. I don't like it for the following reasons:

+ The conversion is meant to designate that the param is immutable, but there are exceptions to this rule (Models are mutable, and params of type Proc are callable. 
+ There is connective friction converting a param `param :my_special_thing` to @MySpecialThing which is unnecessary considering the point above
+ The code is less readable; :my_special_thing != @MySpecialThing, so you have to burn cycles to read it
+ There is Rubu convention with `attr_reader`  where the instance  variable is not converted, but available in the class as `@my_special_thing`
+ More typing! 

So basically, I don't see that we get the advantage (the immutable rule does not hold), but we get the negatives .

My vote is just that the default be reversed: `param :my_special_thign` becomes `@my_special_thing`  and if you want to change it you opt for the behaviour where it becomes `@MySpecialThing`

Anyway, that's what I think. I will go with whatever is decided.

**Comment by catmando** [comment]
@barriehadfield I agree about the params. Have you thought about the other alternative and that is params become accessor methods. so param :my_special_thing creates an accessor method my_special_thing. I personally like this the best for three reasons:

1. it prevents accidental writing to a @my_special_thing which is just going to mess you up when the param gets asyncrounosuly updated.
2. it looks like ruby (i.e. params in ruby are also indistinguishable from methods when you read the code)
3. params can be internally aliased. This allows very nice: param :fuel_pump_on, aka: :fuel_pump_on?

but if everyone feels better keeping them as instance variables I can live with that too.

**Comment by aemadrid** [comment]
The only weird thing for me is the confusion about constants. Uppercase gives me the impression it is a _frozen_ object and that kinda clashes with what I expect from a tag object. Not a big deal though.

**Comment by fkchang** [comment]
If we're going to go against Ruby conventions of lowercase method names and Capital constants, it' needs to be super obvious and repeated, etc... so someone who drops in might not get confused. i prefer lowercased myself going back to markaby, haml, slim, etc. -- that might appeal more to that audience and I can see some people just getting turned off because it's different than standard Ruby conventions

**Comment by catmando** [comment]
Agree that if we continue the way we have it (which seems likely) it needs to be announced at every turn (as I do here: https://youtu.be/GEe7hHIhyUs )

The problem is that haml, slim etc are preprocessors, so there are context markers so that the preprocessor knows the difference between a tag and ruby code.   We could do the same thing.  For example:  _div_ would follow all the rules, make the div noticable, not be likely conflict with any existing names.

But is that better than DIV ?

Because of the name conflicts I really thing just saying i, a, b, p, div, input is a non-starter however.

**Comment by WaKeMaTTa** [comment]
Like @aemadrid said Uppercase words in ruby are reserved for constants. I should say that we follow more the ruby style.

The current code:

```ruby
class HelloWorld < HyperComponent
  render(DIV) do
    # try changing 'world' to your own name
    H1 { "Hello world" }
    P(class: "green-text") { "Let's gets started!" }
  end
end
```

It should be:

```ruby
class HelloWorld < HyperComponent
  render do
    tag :div do
      # try changing 'world' to your own name
      tag :h1 { "Hello world" }
      tag :p, class: "green-text" { "Let's gets started!" }
    end
  end
end
```

:heavy_check_mark: It has a few public methods from HyperStack (only `render` and `tag` in my example).
:heavy_check_mark: It follows the ruby style and ruby community style
:heavy_check_mark: Its visually cleaner and readable
:heavy_check_mark: Its consistent

---

This is how `HyperComponent` should be implemented in the end (more or less):

```ruby
class HyperComponent
  AVAILABLE_TAGs = %i[div h1 p].freeze

  def render
    block_given? ? yield : "" # "" or tag_div
  end

  def tag(name, arguments = {})
    raise(ArgumentError, "First argument '#{name}' is invalid. You should use one of this: #{AVAILABLE_TAGs.join(', ')}") unless AVAILABLE_TAGs.include?(name)

    if block_given?
      send(:"tag_#{name}", arguments) do
        yield
      end
    else
      send(:"tag_#{name}", arguments)
    end
  end

  private

  def tag_div(arguments = {})
    # ...
  end

  def tag_h1(arguments = {})
    # ...
  end

  def tag_p(arguments = {})
    # ...
  end
end
```

**Comment by catmando** [comment]
@WaKeMaTTa You apparently are a very smart guy. :-)

While I am not convinced that I want to type `tag` every time I mention a component, what you suggest is not a bad alternative.

But the nice thing is:  HyperComponent is like ApplicationRecord or ApplicationController .  Hyperstack does not actually define it, you do.  In fact you can name it ApplicationComponent if you want ... i.e. its the application defined base class for all your components.  

So guess what, you can implement your solution today!

```ruby
class HyperComponent 
  ... 
  def tag(name, *args, &block)
     name = name.upcase  # all the tags are in fact already defined as UPCASE constants
     raise(ArgumentError, "#{name} is not a recognized html tag") unless const_defined? name
     send(name, *args, &block)
  end
end
```

This is from memory and untested, but something like that will work fine.

One FYI:

keep in mind that `tag :h1 { "Hello world" }` is not valid ruby syntax.  When using `{ ... }` to define a block you have to wrap any params in parens:  `tag(:h1) { "Hello world" }`

**Comment by Tim-Blokdijk** [comment]
I have no strong opinion about the builtin tag naming. If we stick with `UPERCASED` then it should be clearly explained in the docs so new people don't out right disqualify Hyperstack for failing to follow Ruby conventions. Supporting something along the lines of @WaKeMaTTa `tag` proposal as a default with `UPERCASED` as an option for those who want it might be a good idea.
Long term I would love a dsl like haml that is compatible with Opal but that's far off. Maybe some sort of leaky abstraction to make it more compatible with other js frameworks, no idea how that should work.

But I do fully agree that the `@CamelCase` params are bad for the reasons @barriehadfield mentioned. Using accessor methods should probably be the default.

**Comment by patleb** [comment]
The alternative I used in the past is suffixing tag names with an underscore (which isn't used as any convention that I know of). So, basically, you would get something like this:

```ruby
div_ do
	h1_(class: 'some-css') do
		'some text'
	end
end
```
